File: internal/analyzer/correlation_analyzer.go
package analyzer

import (
	"net-zilla/internal/models"
	"time"
)

type CorrelationAnalyzer struct {
	thresholds CorrelationThresholds
}

type CorrelationThresholds struct {
	TimeWindow       time.Duration
	IPSimilarity     float64
	DomainSimilarity float64
	PatternThreshold int
}

type CorrelationAnalysis struct {
	RelatedAnalyses  []string
	CommonIPs        []string
	CommonDomains    []string
	CommonPatterns   []string
	ClusterScore     int
	ThreatCluster    bool
	Recommendations  []string
}

func NewCorrelationAnalyzer() *CorrelationAnalyzer {
	return &CorrelationAnalyzer{
		thresholds: CorrelationThresholds{
			TimeWindow:       24 * time.Hour,
			IPSimilarity:     0.8,
			DomainSimilarity: 0.7,
			PatternThreshold: 3,
		},
	}
}

func (c *CorrelationAnalyzer) AnalyzeCorrelations(analyses []*models.ThreatAnalysis) *CorrelationAnalysis {
	correlation := &CorrelationAnalysis{}

	if len(analyses) < 2 {
		return correlation
	}

	// Find common IPs
	correlation.CommonIPs = c.findCommonIPs(analyses)

	// Find common domains
	correlation.CommonDomains = c.findCommonDomains(analyses)

	// Find common patterns
	correlation.CommonPatterns = c.findCommonPatterns(analyses)

	// Calculate cluster score
	correlation.ClusterScore = c.calculateClusterScore(correlation)

	// Determine if this is a threat cluster
	correlation.ThreatCluster = c.isThreatCluster(correlation)

	// Generate recommendations
	correlation.Recommendations = c.generateRecommendations(correlation)

	return correlation
}

func (c *CorrelationAnalyzer) findCommonIPs(analyses []*models.ThreatAnalysis) []string {
	ipCount := make(map[string]int)
	
	for _, analysis := range analyses {
		if analysis.IPInfo != nil {
			ipCount[analysis.IPInfo.IP]++
		}
		
		// Check redirect chain IPs
		for _, redirect := range analysis.RedirectChain {
			if redirect.IPAddress != "" {
				ipCount[redirect.IPAddress]++
			}
		}
	}

	var commonIPs []string
	for ip, count := range ipCount {
		if count > 1 {
			commonIPs = append(commonIPs, ip)
		}
	}

	return commonIPs
}

func (c *CorrelationAnalyzer) findCommonDomains(analyses []*models.ThreatAnalysis) []string {
	domainCount := make(map[string]int)
	
	for _, analysis := range analyses {
		domain := extractDomain(analysis.URL)
		if domain != "" {
			domainCount[domain]++
		}
		
		// Check redirect domains
		for _, redirect := range analysis.RedirectChain {
			redirectDomain := extractDomain(redirect.URL)
			if redirectDomain != "" {
				domainCount[redirectDomain]++
			}
		}
	}

	var commonDomains []string
	for domain, count := range domainCount {
		if count > 1 {
			commonDomains = append(commonDomains, domain)
		}
	}

	return commonDomains
}

func (c *CorrelationAnalyzer) findCommonPatterns(analyses []*models.ThreatAnalysis) []string {
	patternCount := make(map[string]int)
	
	for _, analysis := range analyses {
		// Count phishing indicators
		for _, indicator := range analysis.PhishingIndicators {
			patternCount[indicator]++
		}
		
		// Count suspicious features
		for _, feature := range analysis.SuspiciousFeatures {
			patternCount[feature]++
		}
		
		// Count AI-detected threats
		if analysis.AIResult != nil {
			for _, threat := range analysis.AIResult.Threats {
				patternCount[threat]++
			}
		}
	}

	var commonPatterns []string
	for pattern, count := range patternCount {
		if count >= c.thresholds.PatternThreshold {
			commonPatterns = append(commonPatterns, pattern)
		}
	}

	return commonPatterns
}

func (c *CorrelationAnalyzer) calculateClusterScore(correlation *CorrelationAnalysis) int {
	score := 0

	// Score based on common IPs
	score += len(correlation.CommonIPs) * 10

	// Score based on common domains
	score += len(correlation.CommonDomains) * 8

	// Score based on common patterns
	score += len(correlation.CommonPatterns) * 5

	return score
}

func (c *CorrelationAnalyzer) isThreatCluster(correlation *CorrelationAnalysis) bool {
	if correlation.ClusterScore < 20 {
		return false
	}

	// High number of common IPs
	if len(correlation.CommonIPs) >= 2 {
		return true
	}

	// High number of common domains
	if len(correlation.CommonDomains) >= 3 {
		return true
	}

	// Multiple strong patterns
	if len(correlation.CommonPatterns) >= 3 {
		return true
	}

	return false
}

func (c *CorrelationAnalyzer) generateRecommendations(correlation *CorrelationAnalysis) []string {
	var recommendations []string

	if correlation.ThreatCluster {
		recommendations = append(recommendations,
			"🚨 Threat cluster detected - coordinated attack likely",
			"Investigate common IPs and domains for blocklisting",
			"Report coordinated activity to relevant authorities",
		)
	}

	if len(correlation.CommonIPs) > 0 {
		recommendations = append(recommendations,
			"Multiple analyses share common IP addresses",
			"Consider IP-based blocking for repeated offenders",
		)
	}

	if len(correlation.CommonDomains) > 0 {
		recommendations = append(recommendations,
			"Multiple analyses share common domain patterns",
			"Investigate domain registration patterns",
		)
	}

	if len(correlation.CommonPatterns) > 0 {
		recommendations = append(recommendations,
			"Common threat patterns detected across analyses",
			"Update detection rules based on common patterns",
		)
	}

	return recommendations
}

File: netzilla/core/correlation_analyzer.py
from typing import List, Dict
from dataclasses import dataclass, field
from datetime import timedelta
from netzilla.models.threats import ThreatAnalysis

@dataclass
class CorrelationThresholds:
    time_window: timedelta = timedelta(hours=24)
    ip_similarity: float = 0.8
    domain_similarity: float = 0.7
    pattern_threshold: int = 3

@dataclass
class CorrelationAnalysis:
    related_analyses: List[str] = field(default_factory=list)
    common_ips: List[str] = field(default_factory=list)
    common_domains: List[str] = field(default_factory=list)
    common_patterns: List[str] = field(default_factory=list)
    cluster_score: int = 0
    threat_cluster: bool = False
    recommendations: List[str] = field(default_factory=list)

class CorrelationAnalyzer:
    def __init__(self):
        self.thresholds = CorrelationThresholds()

    def analyze_correlations(self, analyses: List[ThreatAnalysis]) -> CorrelationAnalysis:
        correlation = CorrelationAnalysis()

        if len(analyses) < 2:
            return correlation

        correlation.common_ips = self._find_common_ips(analyses)
        correlation.common_domains = self._find_common_domains(analyses)
        correlation.common_patterns = self._find_common_patterns(analyses)
        correlation.cluster_score = self._calculate_cluster_score(correlation)
        correlation.threat_cluster = self._is_threat_cluster(correlation)
        correlation.recommendations = self._generate_recommendations(correlation)

        return correlation

    def _find_common_ips(self, analyses: List[ThreatAnalysis]) -> List[str]:
        ip_count: Dict[str, int] = {}
        for analysis in analyses:
            if analysis.ip_info:
                ip_count[analysis.ip_info.ip] = ip_count.get(analysis.ip_info.ip, 0) + 1
            
            for redirect in analysis.redirect_chain:
                if redirect.ip_address:
                    ip_count[redirect.ip_address] = ip_count.get(redirect.ip_address, 0) + 1
        
        return [ip for ip, count in ip_count.items() if count > 1]

    def _find_common_domains(self, analyses: List[ThreatAnalysis]) -> List[str]:
        domain_count: Dict[str, int] = {}
        from urllib.parse import urlparse

        def extract_domain(url: str) -> str:
            return urlparse(url).hostname or ""

        for analysis in analyses:
            domain = extract_domain(analysis.url)
            if domain:
                domain_count[domain] = domain_count.get(domain, 0) + 1
            
            for redirect in analysis.redirect_chain:
                redirect_domain = extract_domain(redirect.url)
                if redirect_domain:
                    domain_count[redirect_domain] = domain_count.get(redirect_domain, 0) + 1
        
        return [domain for domain, count in domain_count.items() if count > 1]

    def _find_common_patterns(self, analyses: List[ThreatAnalysis]) -> List[str]:
        pattern_count: Dict[str, int] = {}
        for analysis in analyses:
            for indicator in analysis.phishing_indicators:
                pattern_count[indicator] = pattern_count.get(indicator, 0) + 1
            
            for feature in analysis.suspicious_features:
                pattern_count[feature] = pattern_count.get(feature, 0) + 1
            
            if analysis.ai_result:
                for threat in analysis.ai_result.threats:
                    pattern_count[threat] = pattern_count.get(threat, 0) + 1
        
        return [pattern for pattern, count in pattern_count.items() if count >= self.thresholds.pattern_threshold]

    def _calculate_cluster_score(self, correlation: CorrelationAnalysis) -> int:
        score = 0
        score += len(correlation.common_ips) * 10
        score += len(correlation.common_domains) * 8
        score += len(correlation.common_patterns) * 5
        return score

    def _is_threat_cluster(self, correlation: CorrelationAnalysis) -> bool:
        if correlation.cluster_score < 20:
            return False
        if len(correlation.common_ips) >= 2:
            return True
        if len(correlation.common_domains) >= 3:
            return True
        if len(correlation.common_patterns) >= 3:
            return True
        return False

    def _generate_recommendations(self, correlation: CorrelationAnalysis) -> List[str]:
        recommendations = []
        if correlation.threat_cluster:
            recommendations.extend([
                "🚨 Threat cluster detected - coordinated attack likely",
                "Investigate common IPs and domains for blocklisting",
                "Report coordinated activity to relevant authorities",
            ])
        if correlation.common_ips:
            recommendations.extend([
                "Multiple analyses share common IP addresses",
                "Consider IP-based blocking for repeated offenders",
            ])
        if correlation.common_domains:
            recommendations.extend([
                "Multiple analyses share common domain patterns",
                "Investigate domain registration patterns",
            ])
        if correlation.common_patterns:
            recommendations.extend([
                "Common threat patterns detected across analyses",
                "Update detection rules based on common patterns",
            ])
        return recommendations
